staging/src/k8s.io/apiserver/pkg/storage/etcd3/feature_support_checker.go

```diff
@@ -0,0 +1,172 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package etcd3
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"golang.org/x/sync/errgroup"
+	"k8s.io/apimachinery/pkg/util/version"
+	genericfeatures "k8s.io/apiserver/pkg/features"
+	"k8s.io/apiserver/pkg/storage"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+)
+
+var (
+	// This is a error type that we used in isDeprecated
+	ErrDeprecatedCluster = errors.New("")
+	// Define these static versions to use for checking version of etcd, issue on kubernetes #123192
+	v3_4_25 = version.MustParseSemantic("3.4.25")
+	v3_5_0  = version.MustParseSemantic("3.5.0")
+	v3_5_8  = version.MustParseSemantic("3.5.8")
+
+	// RequestWatchProgress is an etcd feature that may use to check if it supported or not.
+	RequestWatchProgress storage.EtcdFeature = "RequestWatchProgress"
+)
+
+// FeatureSupportChecker to define Supports function which checks etcd with assigned feature and return boolean.
+type FeatureSupportChecker interface {
+	Supports(ctx context.Context, feature storage.EtcdFeature) (bool, error)
+}
+
+// FeatureSupportInfo holds information about feature support status and associated error.
+type FeatureSupportInfo struct {
+	Supported bool
+	Err       error
+}
+
+type DefaultFeatureSupportChecker struct {
+	client *clientv3.Client
+	// featureSupportCache store result of feature call with this format: {result: bool, err: error}
+	featureSupportCache map[storage.EtcdFeature]FeatureSupportInfo
+	mutex               sync.RWMutex
```

Comment:
It should be clear what variables are protected by mutex.

Put mutex above the variable it protects, and leave a comment that all variables below are protected by it.